PWR/PWR_RTC/Core/Src/main.c

```diff
@@ -0,0 +1,97 @@
+#include "stm32l053xx.h"
+
+//------ DEFINES
+//--- INPUTS
+#define USR_BT_PRESS		(!(GPIOA->IDR & GPIO_IDR_ID10_Msk))
+
+//--- OUTPUTS
+#define GREEN_LED_ON		(GPIOA->ODR |= GPIO_ODR_OD7_Msk)
+#define	GREEN_LED_OFF		(GPIOA->ODR &= ~GPIO_ODR_OD7_Msk)
+#define TOGGLE_GREEN_LED	(GPIOA->ODR ^=GPIO_IDR_ID7_Msk)
+
+//--- GENERAL
+#define RTC_30_SECONDS	32000	//used for RTC config
+#define RTC_KEY1		0xCA	//1st key to unlock RTC's registers (write to RTC_WRP)
+#define RTC_KEY2		0x53	//2nd key to unlock RTC's registers (write to RTC_WRP)
+
+int main(void)
+{
+	/*							   *
+	 * --- GPIO INITIALIZATION --- *
+	 *							   */
+	/* -> GPIOA
+	 * PA14 -> AF mode (SWD_IO)		| Low speed	| Push pull
+	 * PA13 -> AF mode (SWD_CK)		| Low speed	| Push pull
+	 * PA10	-> Input (User Button)	| Low speed	| Pull-up
+	 */
+	RCC->IOPENR	 |=	RCC_IOPENR_GPIOAEN;		//enable GPIOA clock
+
+	GPIOA->MODER &=	~GPIO_MODER_MODE10_Msk;	//set PA10 as input
+	GPIOA->PUPDR |= GPIO_PUPDR_PUPD10_0;	//enables PA10 pull-up
+
+	//Green led, debug purpose
+	GPIOA->MODER &= ~GPIO_MODER_MODE7_1;	//set PA7 as output
+	GPIOA->BSRR |= GPIO_BSRR_BR_7;			//resets the corresponding OD7 bit
+	GREEN_LED_ON;
+
+	/*									   *
+	 * --- STANDBY MODE INITIALIZATION --- *
+	 *									   */
+	RCC->APB1ENR |= RCC_APB1ENR_PWREN;	//enable PWR clock
+	SCB->SCR = SCB_SCR_SLEEPDEEP_Msk; 	//low-power mode = deep sleep
+	PWR->CR |= PWR_CR_PDDS;				//making sure we're entering standby mode
+	PWR->CR |= PWR_CR_CWUF		|		//clears WUF after 2 system clock cycles
+				PWR_CR_LPSDSR	|		//voltage regulator in low-power mode
+				PWR_CR_ULP		;		//ultra low power mode enable
+
+	//entering standy mode procedure
+	DBGMCU->CR |= DBGMCU_CR_DBG_STOP;	//this bit needs to be set if you're going to debug this code
+
+	/*							   *
+	 *  --- RTC INITIALIZATION --- *
+	 *							   */
+	// --- Resetting RTC registers (debug purpose)
+	RCC->CSR |= RCC_CSR_RTCRST;
+	while(!(RCC->CSR & RCC_CSR_RTCRST));
+	RCC->CSR &= ~RCC_CSR_RTCRST;
+	while((RCC->CSR & RCC_CSR_RTCRST));
+
+	RCC->CSR |= RCC_CSR_LSION;					//LSI oscillator ON
+	while(!(RCC->CSR & RCC_CSR_LSIRDY));		//poll until LSI is stable
+
+	PWR->CR  |=	PWR_CR_DBP;						//enable write access to the RTC and RCC CSR registers
+	RCC->CSR |=	RCC_CSR_RTCSEL_LSI	|			//sets LSI as RTC clock source (37 kHz)
+				RCC_CSR_RTCEN		;			//enables the RTC clock (those bits are write protected!)
+	// --- Unlocking RTC's write protection
+	RTC->WPR = RTC_KEY1;
+	RTC->WPR = RTC_KEY2;
+	// --- RTC's unlocked
+	RTC->CR &= ~RTC_CR_WUTE;					//disables the wakeup timer
+	while(!(RTC->ISR & RTC_ISR_WUTWF));			//polling WUTWF until it is set
+	RTC->ISR &= ~RTC_ISR_WUTF;					//clears RTC wakeup flag
+	RTC->ISR |= RTC_ISR_INIT;					//RTC enters initialization mode
+	while(!(RTC->ISR & RTC_ISR_INITF));			//polling initialization mode flag
+	RTC->PRER = (36 << RTC_PRER_PREDIV_S_Pos);	//set f_apre to 1 kHz

```

Comment:
Why not 1Hz? And why not just RTC->PRER = 36?

Reply:
First thing, it should be **RTC_PRER_PREDIV_A**, not RTC_PRER_PREDIV_S. 
I had some difficulties to understand the Prescaler part in RTC, but I'm using the 7-bit asynchronous Prescaler for the RTC (it looks like to me that the synchronous one is not used for the wakeup event, and, for what I could understand, the asynchronous save more energy than the synchronous). So I couldn't get a 1 Hz output for the asynchronous Prescaler (289 Hz min)

Reply:
You can use both PREDIV_A and PREDIV_S for the auto wakeup, but if you want short timeout periods (up to 32s) you don't need to select the ck_spre as wakeup clock, you can use RTCCLOCK divided by 2, 4, 8 or 16. When using the both PREDIV_A and PREDIV_S you should set PREDIV_A to a high value, in order to minimize consumption, but it doesn't say if one prescaler save more energy than the other.
Anyway, it's ok to use 1kHz now, but I'm no sure about the maximum wakeup time we can get with this wakeup clock, because of the maximum value you can set in the RTC_WUPR register. So maybe for greater times you're gonna have to increase the prescalers.
Nice job until now!

Reply:
So, since I'm setting f_apre to 1 kHz and f_spre (1 Hz), what exactly is happening? Here's where I had some problems to understand all this clocking matter on RTC. I can't remember now why I'm setting them both, but it worked. My RTC clock input is 37 kHz (HSI).

Reply:
Both PREDIV_A and PREDIV_S have default value: PREDIV_A = 0x007F and PREDIV_S = 0x00FF. So if you want to use only the PREDIV_A, you have to set the PREDIV_S = 0x0001, because 1 is the minimum value. Read the page 647 from the RM to understand better how the prescaler work, but basically, when using ck_spre as wakeup clock, the value of ck_spre will be the RTC clock (LSI in this case) divided by PREDIV_A and then divided by PREDIV_S, or RTC clock / [(PREDIV_A +1)*(PREDIV_S+1)].

And you're using LSI right?

Reply:
Yes! LSI!

I'll check that again.